src/Common/src/Interop/Unix/System.Native/Interop.ReadLink.cs

```diff
@@ -16,9 +18,47 @@ internal static partial class Sys
         /// <param name="buffer">The buffer to hold the output path</param>
         /// <param name="bufferSize">The size of the buffer</param>
         /// <returns>
-        /// Returns the number of bytes placed into the buffer on success; otherwise, -1 is returned
+        /// Returns the number of bytes placed into the buffer on success; 0 if the buffer is too small; and -1 on error.
         /// </returns>
         [DllImport(Libraries.SystemNative, EntryPoint = "SystemNative_ReadLink", SetLastError = true)]
-        internal static extern unsafe int ReadLink(string path, byte[] buffer, int bufferSize);
+        private static extern unsafe int ReadLink(string path, byte[] buffer, int bufferSize);
+
+        /// <summary>
+        /// Takes a path to a symbolic link and returns the link target path.
+        /// </summary>
+        /// <param name="path">The path to the symlink</param>
+        /// <returns>
+        /// Returns the link to the target path on success; and null otherwise.
+        /// </returns>
+        public static string ReadLink(string path)
+        {
+            int bufferSize = 256;
+            do
+            {
+                byte[] buffer = ArrayPool<byte>.Shared.Rent(bufferSize);
+                try
+                {
+                    int resultLength = Interop.Sys.ReadLink(path, buffer, buffer.Length - 1); // -1 for null termination
+                    if (resultLength > 0)
+                    {
+                        // success, null terminate
+                        buffer[resultLength] = (byte)'\0';
+                        return Encoding.UTF8.GetString(buffer, 0, resultLength);
+                    }
+                    else if (resultLength < 0)
+                    {
+                        // error
+                        return null;
+                    }
+                }
+                finally
+                {
+                    ArrayPool<byte>.Shared.Return(buffer);
+                }
+
+                // buffer was too small, loop around again and try with a larger buffer.
+                bufferSize *= 2;
```

Comment:
We're now growing unbounded?

Reply:
We're growing the buffer until the native function returns successful/fails.

Reply:
Is it possible that might never happen, e.g. with a recursive symlink or something?

Reply:
We're only allocating more when there is insufficient space for the return path.
readlink returns ELOOP when it has to deal with too many symbolic links in the path argument.
You need to give it a link and then you learn where that one points to. That path isn't followed, it is returned.

Reply:
Ok